test/e2e/certificate_publisher_test.go

```diff
@@ -0,0 +1,168 @@
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	ingresscontroller "github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// TestCreateIngressControllerThenSecret creates an ingresscontroller that
+// references a secret that does not exist, then creates the secret and verifies
+// that the operator updates the "router-certs" global secret.
+func TestCreateIngressControllerThenSecret(t *testing.T) {
+	cl, ns, err := getClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create the ingresscontroller.
+	var one int32 = 1
+	ic := &operatorv1.IngressController{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-ingresscontroller-then-secret",
+			Namespace: ns,
+		},
+		Spec: operatorv1.IngressControllerSpec{
+			DefaultCertificate: &corev1.LocalObjectReference{
+				Name: "e2e-test-secret",
+			},
+			Domain: "test-ingresscontroller-then-secret",
+			EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
+				Type: operatorv1.PrivateStrategyType,
+			},
+			Replicas: &one,
+		},
+	}
+	if err := cl.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create the ingresscontroller: %v", err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), ic); err != nil {
+			t.Fatalf("failed to delete the ingresscontroller: %v", err)
+		}
+	}()
+
+	// Wait for the deployment to exist.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
```

Comment:
How does polling for the deployment compare to polling for the ingresscontroller to report an available condition (or availableReplicas prior to our population of conditions)? I see these deployment checks elsewhere and wonder why we're using them

Reply:
And why is an availability check of any kind needed at all here?

Reply:
Polling for the `availableReplicas` would mean the test would fail if the deployment does not become available, which is not strictly part of what it is testing.  You're right that we could do without the poll, although I would like to have something to ensure the controllers process the two events (creation of the ingresscontroller and creation of the secret) in that specific order, since that is the purpose of the test.

Reply:
Changed to poll the ingresscontroller and check that an effective domain has been set.